astropy/io/ascii/core.py

```diff
@@ -1092,13 +1092,18 @@ def _convert_vals(self, cols):
                     col.converters.pop(0)
                     last_err = err
                 except OverflowError as err:
-                    # Overflow during conversion (most likely an int that
-                    # doesn't fit in native C long). Put string at the top of
-                    # the converters list for the next while iteration.
-                    warnings.warn(
-                        "OverflowError converting to {} in column {}, reverting to String."
-                        .format(converter_type.__name__, col.name), AstropyWarning)
-                    col.converters.insert(0, convert_numpy(numpy.str))
+                        # Overflow during conversion (most likely an int that
+                        # doesn't fit in native C long).
+                    if converter_type == IntType and not isinstance(last_err, OverflowError):
+                        # First attempt: use an int64, which may help on 32-bit platforms
+                        col.converters[0] = convert_numpy(numpy.int64)
```

Comment:
I did this first, but made some minor error. This lead to an infinite loop here which was hard to discover -- a new type was inserter here, and then silently popped back out here: https://github.com/astropy/astropy/blob/3acc889f0372abf28d6f4688b542567018dcf911/astropy/io/ascii/core.py#L1091-L1093

Then it went into the next iteration, to again raise the overflow error. I found that a bit surprising and it took me a while to discover. I think this insert/pop procedure is not very robust here, and it is better to have the code definitely completing in finite time. If we already know that a certain type causes an overflow error, we can just replace it.

Reply:
Good point. This suggests to me a different implementation that might work out better which is roughly to change
`default_converters` to:
```
    default_converters = [
        convert_numpy(int),
        convert_numpy(np.int64),
        convert_numpy(float),
        convert_numpy(str)]
```
This doesn't exactly work with the existing logic, but the advantage here is reducing the special case logic in the code. Those converters are also something the user can specify, so reducing the magic in the code would be good to prevent surprises. There are already problems, e.g. if the user says the type of a column should be `int`, I think the current code will automatically drop through to `str` for an overflow with a warning, but maybe that should be an exception if `str` is not in the original list of converters.

Agreed this code is worth improving a bit. I'll think about this a bit more if you don't come up with something first.

Reply:
I agree here that just adding `np.int64` to the default converters (and removing the special code for `MemoryOverflow`) would help. I was just not sure then about the special cases: a very long sequence of digits may overflow the int64 and would currently direct fall through the `str`, while with this change it may be converted to `float`. I am not sure what is intended here.